### PR TITLE
refactor service parameters as context

### PR DESCRIPTION
**Pull request by michaelmattig**
makes service parameters more uniform and more easily extensible

### CODE REVIEW

File: services/src/handlers/mod.rs [hunk 1]
```diff
@@ -32,20 +38,68 @@ pub async fn handle_rejection(error: Rejection) -> Result<impl Reply, Rejection>
...
+        self.user_db.clone()
+    }
+
+    fn project_db(&self) -> DB<Self::ProjectDB> {
+        self.project_db.clone()
+    }
+
+    fn workflow_registry(&self) -> DB<Self::WorkflowRegistry> {
+        self.workflow_registry.clone()
+    }
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/mod.rs:94]: Is it necessary to always clone it or to hand out a reference for some cases?

File: services/src/handlers/mod.rs [hunk 2]
```diff
@@ -32,20 +38,68 @@ pub async fn handle_rejection(error: Rejection) -> Result<impl Reply, Rejection>
...
     }
 
     warp::any()
-        .and(warp::any().map(move || Arc::clone(&user_db)))
+        .and(warp::any().map(move || ctx.clone()))
         .and(warp::header::<String>("authorization"))
         .and_then(do_authenticate)
 }
+
+pub trait Context: 'static + Send + Sync + Clone {
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/mod.rs:58]: I recommend writing some doc-comments.

File: services/src/handlers/mod.rs [hunk 3]
```diff
@@ -32,20 +38,68 @@ pub async fn handle_rejection(error: Rejection) -> Result<impl Reply, Rejection>
...
+    fn user_db(&self) -> DB<Self::UserDB>;
+    fn project_db(&self) -> DB<Self::ProjectDB>;
+    fn workflow_registry(&self) -> DB<Self::WorkflowRegistry>;
+    fn session(&self) -> Result<&Session>;
+
+    fn set_session(&mut self, session: Session);
+}
+
+#[derive(Clone, Default)]
+pub struct DefaultContext {
```
Comment (ChristianBeilschmidt) [inline services/src/handlers/mod.rs:72]:
```suggestion
pub struct InMemoryContext {
```

**Review by ChristianBeilschmidt — CHANGES_REQUESTED** [review]

File: services/src/handlers/mod.rs (hunk 1 above)
Comment (michaelmattig) [reply services/src/handlers/mod.rs:94]: Arc is already kind of a reference. Does it really make sense to hand out a reference to a reference? Maybe it makes more sense to (additionally) provide a method that directly returns the contained Mutex or even the Read/Write guard. Latter one would introduce two methods for each contained value for read/write...

File: services/src/handlers/mod.rs (hunk 2 above)
Comment (michaelmattig) [reply services/src/handlers/mod.rs:60]: 3a6bd99fc4447b0eaec36c53dfa09c76ec330398

File: services/src/handlers/mod.rs (hunk 3 above)
Comment (michaelmattig) [reply services/src/handlers/mod.rs]: 3a6bd99fc4447b0eaec36c53dfa09c76ec330398

File: services/src/handlers/mod.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply services/src/handlers/mod.rs:97]: I see lots of calls where it is not necessary to clone it.
So you could hand out some read guard with `project_db_ref()` and write guard with `project_db_mut()`.

Comment (michaelmattig) [reply services/src/handlers/mod.rs]: c6abcb51f4bd8e4afa329878a50068a4b73ec59f

**Pull request merged**